### PR TITLE
DBZ-3621 Saving Debezium offsets in JDBC database

### PR DESCRIPTION
**Pull request by subkanthi**
Added debezium-storage-jdbc. 
@jpechane , Creating a new PR as discussed in #4312. Thanks

### CODE REVIEW

**Comment by subkanthi** [comment]
> @subkanthi Thanks, could you please try to address the other comment in the related PRs?

Hi @jpechane , I went through this PR https://github.com/debezium/debezium/pull/2543 and addressed the missing ones related to comments. Please let me know if you find something that is not addressed.

**Comment by jpechane** [comment]
@subodh1810 I am sorry, probably some were lost :-(. In general I'd prefer to have at least CREAT nd SELECT queries configurable. The reasoning behind it that they might be optimized by using better dtataypes, queries on different databases.

Also could you please use naming JDBC URL, not URI? IMHO URL is definitely more often used.

**Comment by jpechane** [comment]
@subkanthi I don't think it is necessary. It owuld be just enought to properly document each query. IMHO people will mostly change the datatypes but will keep the column names same.

File: debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java [hunk 1]
```diff
@@ -265,6 +265,32 @@ public final class EmbeddedEngine implements DebeziumEngine<SourceRecord> {
...
+     * record_insert_ts - Timestamp when the record was inserted
+     * record_insert_seq - Sequence number of record
+     */
+    public static final Field OFFSET_STORAGE_TABLE_DDL = Field.create("offset.storage.jdbc.offset_table_ddl")
+            .withDescription("Create table syntax for offset jdbc table")
+            .withDefault(DEFAULT_OFFSET_STORAGE_TABLE_DDL);
+
+    public static final String DEFAULT_OFFSET_STORAGE_TABLE_SELECT = "SELECT id, offset_key, offset_val FROM %s " +
+            "ORDER BY record_insert_ts, record_insert_seq";
+    public static final Field OFFSET_STORAGE_TABLE_SELECT = Field.create("offset.storage.jdbc.offset_table_select`")
```
Comment (jpechane) [inline debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java:290]:
```suggestion
    public static final Field OFFSET_STORAGE_TABLE_SELECT = Field.create("offset.storage.jdbc.offset_table_select")
```

File: debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java [hunk 2]
```diff
@@ -265,6 +265,32 @@ public final class EmbeddedEngine implements DebeziumEngine<SourceRecord> {
...
+    /**
+     * The JDBC table that will store offset information.
+     * id - UUID
+     * offset_key - Offset Key
+     * offset_val - Offset value
+     * record_insert_ts - Timestamp when the record was inserted
+     * record_insert_seq - Sequence number of record
+     */
+    public static final Field OFFSET_STORAGE_TABLE_DDL = Field.create("offset.storage.jdbc.offset_table_ddl")
+            .withDescription("Create table syntax for offset jdbc table")
```
Comment (jpechane) [inline debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java:285]:
```suggestion
            .withDescription("Create table syntax for offset JDBC table")
```

File: debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java [hunk 3]
```diff
@@ -265,6 +265,32 @@ public final class EmbeddedEngine implements DebeziumEngine<SourceRecord> {
...
+            .withDefault(DEFAULT_OFFSET_STORAGE_TABLE_DDL);
+
+    public static final String DEFAULT_OFFSET_STORAGE_TABLE_SELECT = "SELECT id, offset_key, offset_val FROM %s " +
+            "ORDER BY record_insert_ts, record_insert_seq";
+    public static final Field OFFSET_STORAGE_TABLE_SELECT = Field.create("offset.storage.jdbc.offset_table_select`")
+            .withDescription("Select syntax to get offset data from jdbc table")
```
Comment (jpechane) [inline debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java:291]:
```suggestion
            .withDescription("Select syntax to get offset data from JDBC table")
```

File: debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java [hunk 4]
```diff
@@ -277,7 +303,8 @@ public final class EmbeddedEngine implements DebeziumEngine<SourceRecord> {
             OFFSET_FLUSH_INTERVAL_MS, OFFSET_COMMIT_TIMEOUT_MS,
             ERRORS_MAX_RETRIES, ERRORS_RETRY_DELAY_INITIAL_MS, ERRORS_RETRY_DELAY_MAX_MS,
             // JDBC OFFSET STORAGE FIELDS
-            OFFSET_STORAGE_JDBC_URL, OFFSET_STORAGE_JDBC_USER, OFFSET_STORAGE_JDBC_PASSWORD, OFFSET_STORAGE_JDBC_TABLE_NAME);
+            OFFSET_STORAGE_JDBC_URL, OFFSET_STORAGE_JDBC_USER, OFFSET_STORAGE_JDBC_PASSWORD, OFFSET_STORAGE_JDBC_TABLE_NAME,
+            OFFSET_STORAGE_TABLE_DDL, OFFSET_STORAGE_TABLE_SELECT);
```
Comment (jpechane) [inline debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java:307]: Please don't add the fields to the config. See `RedisOffsetBackingStore`

File: debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java [hunk 5]
```diff
@@ -1234,6 +1261,8 @@ protected static class EmbeddedConfig extends WorkerConfig {
             Field.group(config, "jdbc", OFFSET_STORAGE_JDBC_USER);
             Field.group(config, "jdbc", OFFSET_STORAGE_JDBC_PASSWORD);
             Field.group(config, "jdbc", OFFSET_STORAGE_JDBC_TABLE_NAME);
+            Field.group(config, "jdbc", OFFSET_STORAGE_TABLE_DDL);
```
Comment (jpechane) [inline debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java:1264]: Same as above

File: debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConfig.java [hunk 6]
```diff
@@ -35,7 +37,15 @@ public class JdbcConfig extends WorkerConfig {
                 .define(OFFSET_STORAGE_TABLE_NAME.name(),
                         ConfigDef.Type.STRING,
                         ConfigDef.Importance.HIGH,
-                        "Name of the table to store offsets");
+                        "Name of the table to store offsets")
+                .define(OFFSET_STORAGE_TABLE_DDL.name(),
```
Comment (jpechane) [inline debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConfig.java:41]: Should not be needed here.

**Review by jpechane — COMMENTED** [review]
@subkanthi We are getting there, thanks! Could you please just revisit the config a bit more? Check the class I pointed too.

File: debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java (hunk 4 above)
Comment (subkanthi) [reply debezium-embedded/src/main/java/io/debezium/embedded/EmbeddedEngine.java:307]: Hi @jpechane , this was added here for the `AbstractConnectorTest`, I also checked with Chris Cranford, its duplicated unfortunately, I checked `RedisOffsetBackingStore`, I dont see an Integration test, the part I can do is to separate the config to a separate file, so its cleaner.

File: debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConfig.java (hunk 6 above)
Comment (subkanthi) [reply debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConfig.java]: WorkerConfig seems to take ConfigDef and Properties as parameters, not sure what needs to be done here.

**Comment by jpechane** [comment]
@subkanthi I rebased the PR on the latest main and made significant changes in the way how the storages are configured.
Please take a look at class `JdbcSchemaHistoryConfig` - we need to make DDL and both SELECTs configurable and then we are good to go and the PR can be merged.

**Comment by subkanthi** [comment]
> @subkanthi I rebased the PR on the latest main and made significant changes in the way how the storages are configured. Please take a look at class `JdbcSchemaHistoryConfig` - we need to make DDL and both SELECTs configurable and then we are good to go and the PR can be merged.

Hi @jpechane , I have added the changes to make SELECT and DDL configurable for `JdbcSchemaHistoryConfig`.

**Comment by jpechane** [comment]
@subkanthi And we are done. Thanks a lot!

**Pull request merged**